nexus/src/db/lookup.rs

```diff
@@ -0,0 +1,754 @@
+    db::model::Name,
+};
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::{ExpressionMethods, QueryDsl, SelectableHelper};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use omicron_common::api::external::{LookupResult, LookupType, ResourceType};
+use uuid::Uuid;
+
+// TODO-dap XXX Neither "fetcH" nor "lookup" needs to be a trait now that the
+// macro is defining the impls and the structs
+
+pub trait Fetch {
+    type FetchType;
+    fn fetch(&self) -> BoxFuture<'_, LookupResult<Self::FetchType>>;
+}
+
+// This private module exist solely to implement the "Sealed trait" pattern.
+// This isn't about future-proofing ourselves.  Rather, we don't want to expose
+// an interface that accesses database objects without an authz check.
+mod private {
+    use super::LookupPath;
+    use futures::future::BoxFuture;
+    use omicron_common::api::external::LookupResult;
+
+    // TODO-dap XXX-dap we could probably get rid of this trait and its impls
+    // because we're now calling `lookup_root()` from a macro that can
+    // essentially inline the impl of GetLookupRoot for Key.
+    pub trait GetLookupRoot {
+        fn lookup_root(&self) -> &LookupPath<'_>;
+    }
+
+    pub trait LookupNoauthz {
+        type LookupType;
+        fn lookup(&self) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+    }
+}
+
+use private::GetLookupRoot;
+use private::LookupNoauthz;
+
+pub trait LookupFor {
+    type LookupType;
+    fn lookup_for(
+        &self,
+        action: authz::Action,
+    ) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+}
+
+enum Key<'a, P> {
+    Name(P, &'a Name),
+    Id(LookupPath<'a>, Uuid),
+}
+
+impl<'a, T> GetLookupRoot for Key<'a, T>
+where
+    T: GetLookupRoot,
+{
+    fn lookup_root(&self) -> &LookupPath<'_> {
+        match self {
+            Key::Name(parent, _) => parent.lookup_root(),
+            Key::Id(lookup, _) => lookup,
+        }
+    }
+}
+
+impl<'a, P> Key<'a, P> {
+    fn lookup_type(&self) -> LookupType {
+        match self {
+            Key::Name(_, name) => LookupType::ByName(name.as_str().to_string()),
+            Key::Id(_, id) => LookupType::ById(*id),
+        }
+    }
+}
+
+pub struct LookupPath<'a> {
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+}
+
+impl<'a> LookupPath<'a> {
+    pub fn new<'b, 'c>(
+        opctx: &'b OpContext,
+        datastore: &'c DataStore,
+    ) -> LookupPath<'a>
+    where
+        'b: 'a,
+        'c: 'a,
+    {
+        LookupPath { opctx, datastore }
+    }
+
+    pub fn organization_name<'b, 'c>(self, name: &'b Name) -> Organization<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Organization { key: Key::Name(self, name) }
+    }
+
+    pub fn organization_id(self, id: Uuid) -> Organization<'a> {
+        Organization { key: Key::Id(self, id) }
+    }
+
+    pub fn project_id(self, id: Uuid) -> Project<'a> {
+        Project { key: Key::Id(self, id) }
+    }
+
+    pub fn instance_id(self, id: Uuid) -> Instance<'a> {
```

Comment:
I know this needs to be here because it can be called directly on a `LookupPath`, but it should also be on the `impl Project`, right?

Reply:
I don't _think_ so -- but I could be wrong.  I believe in the cases where we start with an instance id today, I think we want to operate on it regardless of what Project it's in.  If a use case does care what Project it's in, you can fetch it from `LookupPath::instance_id` and then check the project_id.

The question that sent me down this path is: if you used `Project::instance_id()` to fetch an Instance that exists but is in a different Project, would that be returned or would you get a 404?  I can see someone being surprised by either answer.  ("I just looked this up by instance_id and I can see that it's there -- how did it not find it?" vs. "why did this return something that's not in the Project I was looking in?")

I don't feel strongly about it.  All things being equal I like to defer it.